Tests/NIOSOCKSTests/SocksClientHandler+Tests.swift

```diff
@@ -230,5 +235,48 @@ class SocksClientHandlerTests: XCTestCase {
         XCTAssertNoThrow(self.channel.pipeline.addHandler(handler))
         self.assertOutputBuffer([0x05, 0x01, 0x00])
     }
+    
+    func testDelayedHandlerRemoved() {
+        // for this test we can't use the created channel and handler by setUp
+        self.channel = EmbeddedChannel()
+        let promise = channel.eventLoop.makePromise(of: Void.self)
+        self.handler = SOCKSClientHandler(targetAddress: .address(try! .init(ipAddress: "192.168.1.1", port: 80))) {
+            channel -> EventLoopFuture<Void> in
+            
+            return promise.futureResult
+        }
+        
+        XCTAssertNoThrow(try self.channel.pipeline.syncOperations.addHandler(self.handler))
+        self.connect()
+        
+        // the client should start the handshake instantly
+        self.assertOutputBuffer([0x05, 0x01, 0x00])
+        
+        // server selects an authentication method
+        self.writeInbound([0x05, 0x00])
+        
+        // client sends the request
+        self.assertOutputBuffer([0x05, 0x01, 0x00, 0x01, 192, 168, 1, 1, 0x00, 0x50])
+        
+        // server replies yay
+        self.writeInbound([0x05, 0x00, 0x00, 0x01, 192, 168, 1, 1, 0x00, 0x50])
+        
+        // the SOCKS handler remains in the channel pipeline until we succeed or fail the promise
+        XCTAssertNoThrow(try self.channel.pipeline.syncOperations.handler(type: SOCKSClientHandler.self))
```

Comment:
Can we check that the data still goes straight through the handler before it's removed? That's very important, otherwise we might lose bytes!

Reply:
Not applicable anymore